p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -0,0 +1,396 @@
+			if err == io.EOF {
+				break
+			}
+			return nil, err
+		}
+		if err := n.Push(share[:namespaceSize], share[namespaceSize:]); err != nil {
+			return nil, err
+		}
+	}
+	// to trigger the collection of nodes:
+	_ = n.Root()
+	return nodes, nil
+}
+
+func prependNode(newNode node.Node, nodes *[]node.Node) {
+	*nodes = append(*nodes, node.Node(nil))
+	copy((*nodes)[1:], *nodes)
+	(*nodes)[0] = newNode
+}
+
+func NmtNodeParser(block blocks.Block) (node.Node, error) {
+	const (
+		// length of the domain separator for leaf and inner nodes:
+		prefixOffset = 1
+		// nmtHashSize = flagSize+sha256.Size
+		nmtHashSize = namespaceSize*2 + sha256.Size
+	)
+	var (
+		leafPrefix  = []byte{nmt.LeafPrefix}
+		innerPrefix = []byte{nmt.NodePrefix}
+	)
+	data := block.RawData()
+	if len(data) == 0 {
+		return &nmtLeafNode{
+			cid:  cid.Undef,
+			Data: nil,
+		}, nil
+	}
+	domainSeparator := data[:prefixOffset]
+	if bytes.Equal(domainSeparator, leafPrefix) {
+		return &nmtLeafNode{
+			cid:  block.Cid(),
+			Data: data[prefixOffset:],
+		}, nil
+	}
+	if bytes.Equal(domainSeparator, innerPrefix) {
+		return nmtNode{
+			cid: block.Cid(),
+			l:   data[prefixOffset : prefixOffset+nmtHashSize],
+			r:   data[prefixOffset+nmtHashSize:],
+		}, nil
+	}
+	return nil, fmt.Errorf(
+		"expected first byte of block to be either the leaf or inner node prefix: (%x, %x), got: %x)",
+		leafPrefix,
+		innerPrefix,
+		domainSeparator,
+	)
+}
+
+var _ node.Node = (*nmtNode)(nil)
+var _ node.Node = (*nmtLeafNode)(nil)
+
+type nmtNode struct {
+	// TODO(ismail): we might want to export these later
+	cid  cid.Cid
+	l, r []byte
+}
+
+func (n nmtNode) RawData() []byte {
+	return append([]byte{nmt.NodePrefix}, append(n.l, n.r...)...)
+}
+
+func (n nmtNode) Cid() cid.Cid {
+	return n.cid
+}
+
+func (n nmtNode) String() string {
+	return fmt.Sprintf(`
+node {
+	hash: %x,
+	l: %x,
+	r: %x"
+}`, n.cid.Hash(), n.l, n.r)
+}
+
+func (n nmtNode) Loggable() map[string]interface{} {
+	return nil
+}
+
+func (n nmtNode) Resolve(path []string) (interface{}, []string, error) {
+	switch path[0] {
+	case "0":
+		left, err := cidFromNamespacedSha256(n.l)
+		if err != nil {
+			return nil, nil, err
+		}
+		return &node.Link{Cid: left}, path[1:], nil
+	case "1":
+		right, err := cidFromNamespacedSha256(n.r)
+		if err != nil {
+			return nil, nil, err
+		}
+		return &node.Link{Cid: right}, path[1:], nil
+	default:
+		return nil, nil, errors.New("invalid path for inner node")
+	}
+}
+
+func (n nmtNode) Tree(path string, depth int) []string {
+	if path != "" || depth == 0 {
+		return nil
+	}
+
+	return []string{
+		"0",
+		"1",
+	}
+}
+
+func (n nmtNode) ResolveLink(path []string) (*node.Link, []string, error) {
+	obj, rest, err := n.Resolve(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	lnk, ok := obj.(*node.Link)
+	if !ok {
+		return nil, nil, errors.New("was not a link")
+	}
+
+	return lnk, rest, nil
+}
+
+func (n nmtNode) Copy() node.Node {
+	panic("implement me")
+}
```

Comment:
Can we get a TODO comment here so it's easily searchable / revealed by most IDEs?

Reply:
I just implemented it in 91e9af4.